src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java

```diff
@@ -1,99 +1,327 @@
+        Map<GroupMetaInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>> histogramsOfAllReads  = accumLevels.contains(MetricAccumulationLevel.ALL_READS) ? new HashMap<>() : null;
+        aggregateHistograms(histogramsOfReadGroups, histogramsOfLibraries, histogramsOfSamples, histogramsOfAllReads);
+
+        // convert to htsjdk Histogram and compute metrics
+        if(accumLevels.contains(MetricAccumulationLevel.READ_GROUP)) {
+            statsOfReadGroups  = new HashMap<>();
+            convertSortedMapToHTSHistogram(histogramsOfReadGroups, statsOfReadGroups, histogramMADTolerance);
+        }
 
-        metrics.READ_PAIRS = filteredReads.count();
+        if(accumLevels.contains(MetricAccumulationLevel.LIBRARY)) {
+            statsOfLibraries  = new HashMap<>();
+            convertSortedMapToHTSHistogram(histogramsOfLibraries, statsOfLibraries, histogramMADTolerance);
+        }
+        if(accumLevels.contains(MetricAccumulationLevel.SAMPLE)) {
+            statsOfSamples    = new HashMap<>();
+            convertSortedMapToHTSHistogram(histogramsOfSamples, statsOfLibraries, histogramMADTolerance);
+        }
+        if(accumLevels.contains(MetricAccumulationLevel.ALL_READS)) {
+            statsOfAllReads   = new HashMap<>();
+            convertSortedMapToHTSHistogram(histogramsOfAllReads, statsOfAllReads, histogramMADTolerance);
+        }
+    }
 
-        collectLevelInfo();
+    /**
+     * Utility getter for retrieving stats info for a particular group, given its name.
+     * Stats info returned are organized by pair orientations.
+     * If a particular pair orientation is unavailable (i.e. no reads of this group has pairs of that orientation), it is not returned.
+     * Throws IllegalArgumentException if the requested groupName doesn't exist at the requested level.
+     * @param groupName  String representation of the group's name/id, whose stats information is requested.
+     * @return           the requested group's stats information, organized by pair orientations
+     * @throws           IllegalArgumentException
+     */
+    public Map<SamPairUtil.PairOrientation, Tuple2<Histogram<Integer>, InsertSizeMetrics>> getHistogramsAndMetrics(final String groupName,
+                                                                                                                   final MetricAccumulationLevel level)
+            throws IllegalArgumentException {
+
+        Map<SamPairUtil.PairOrientation, Tuple2<Histogram<Integer>, InsertSizeMetrics>> result = null;
+        switch (level){
+            case ALL_READS:
+                result = statsOfAllReads.get(new GroupMetaInfo(null, null, null, MetricAccumulationLevel.ALL_READS));
+            break;
+            case SAMPLE:
+                for(final GroupMetaInfo groupMetaInfo : statsOfSamples.keySet()){
```

Comment:
This looks like a bit of code duplication across these cases. It seems like you extract the body of these switch cases into another method, or just do the work after the switch statement, and simply pick out the map you want within the switch.